docs/current_docs/manuals/developer/functions.mdx

```diff
@@ -338,6 +338,109 @@ The result will look like this:
 Hello, world
 ```
 
+## Contextual values
+
+Function arguments can define a default contextual value if no value is supplied for them. This default value will be load prior to the function call which makes it more cache efficient.
```

Comment:
I expect that for a new user, it would not be clear how "contextual values" are different from the "default values" explained above. Is a "contextual value" equivalent to adding a "default value for a dir/file" type, or is there something else to know here? 

If different I would like us to add an explanation for WHY this difference exists and also specify HOW the function behaviour is different in each case, also some advice on which to use when (although this may be self-explanatory once the previous items are explained).

Reply:
A dir or file can have a default value (in Typescript & Python, not go)
So it's a bit different, it's a default value based on the context of the module.

I think I already gave an explanation just below in the note

```
:::note
Contextual values are only available for arguments of type `Directory` and `File` for now.

It's specially useful for static directory or file provided to the module. For example, a codebase source's code.
:::
```

Reply:
> A dir or file can have a default value (in Typescript & Python, not go)

Tangential: We don't actually mention this anywhere in the current docs, it seems important to add this limitation in the "default values" section.

TBH I'm still confused by the above explanation. Assume I have a module in Go/Python/TS which accepts a dir argument. 
- Can I define both a default value and a contextual value in Python/TS?

    ```
    # is this valid?
    repo: Annotated[dagger.Directory = "/home", DefaultPath("/")]    
    ```

- If I define both default and contextual value for a Go function, would Go simply ignore the default value, or throw an error, or ..?
- Since Python and TS support both, if I define both, which one would win?

Sorry if these are stupid questions, I'm trying to understand it better so I can help make the explanation clearer!

Reply:
The default value would win, that's not a stupid question at all and it's actually a case covered in the code!
I'll add some explanation around that